ethcore/src/client/client.rs

```diff
@@ -327,23 +327,25 @@ impl Importer {
 		};
 
 		{
-			if !imported_blocks.is_empty() && is_empty {
+			if !imported_blocks.is_empty() {
 				let route = ChainRoute::from(import_results.as_ref());
 
 				if is_empty {
 					self.miner.chain_new_blocks(client, &imported_blocks, &invalid_blocks, route.enacted(), route.retracted(), false);
 				}
 
-				client.notify(|notify| {
-					notify.new_blocks(
-						imported_blocks.clone(),
-						invalid_blocks.clone(),
-						route.clone(),
-						Vec::new(),
-						proposed_blocks.clone(),
-						duration,
-					);
-				});
+				if client.queue_info().total_queue_size() < 10 {
```

Comment:
Not sure whether calling `queue_info` here is expensive. We fetch many other not-used info by this and it acquires three atomics. And I suppose this function is something that needs to be really fast, so it may matter.

Reply:
You're right that it includes a bunch of stuff that we don't need. It currently requires 3 Mutex locks  - I have a separate suggestion for a way to negate the need to lock [WIP](https://github.com/paritytech/parity-ethereum/pull/9988), which could mitigate risk here.